### PR TITLE
Cylinder resource and collision shape (bullet only)

### PR DESCRIPTION
**Pull request by muiroc**
looks like it works...

some notes and potential issues:
- bullet only, it crashes with SW, maybe I should at least add a warning message
- doesnt expose bullet cylinder's half extent constructor, but only the radius and height
- can the PhysicsServer::ShapeType values change or must the existing order be preserved?
- symmetry axis is up, in sync with the mesh cylinder, not with the current capsule shape (99% of use cases imho)
- no docs
- icon most likely must be fixed

### CODE REVIEW

**Comment by groud** [comment]
Related to #17252

File: editor/icons/icon_cylinder_shape.svg [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg width="16" height="16" version="1.1" viewBox="0 0 14.999999 14.999999" xmlns="http://www.w3.org/2000/svg">
```
Comment (aaronfranke) [inline editor/icons/icon_cylinder_shape.svg]: Why "14.999999" instead of "15"?

Comment (muiroc) [reply editor/icons/icon_cylinder_shape.svg]: Indeed as mentioned I should fix the icon, I hastily traced it in inkscape from either icon_c_s_g_cylinder.svg or icon_cylinder_mesh.svg then cleaned it manually, but those where paths! should this three shapes also be converted to paths?
Can i just keep the separated shapes and round the values?

I was thinking of fixing that if the rest of the PR was ok and acceptable

File: modules/bullet/shape_bullet.cpp [hunk 2]
```diff
@@ -113,6 +113,10 @@ btCapsuleShapeZ *ShapeBullet::create_shape_capsule(btScalar radius, btScalar hei
 	return bulletnew(btCapsuleShapeZ(radius, height));
 }
 
+btCylinderShape *ShapeBullet::create_shape_cylinder(btScalar radius, btScalar height) {
+	return bulletnew(btCylinderShape(btVector3(radius, height / btScalar(2), radius)));
```
Comment (AndreaCatania) [inline modules/bullet/shape_bullet.cpp]: is not required to use `btScalar(2)`, you can just use 2.0

**Review by AndreaCatania — CHANGES_REQUESTED** [review]
Before I'll check the code some things should be fixed.

> - bullet only, it crashes with SW, maybe I should at least add a warning message

This must never happen, so you need address it.

> - doesnt expose bullet cylinder's half extent constructor, but only the radius and height

Correct

> - can the PhysicsServer::ShapeType values change or must the existing order be preserved?

Is fine like this.

> - symmetry axis is up, in sync with the mesh cylinder, not with the current capsule shape (99% of use cases imho)

Perfect.

File: modules/bullet/shape_bullet.cpp (hunk 2 above)
Comment (muiroc) [reply modules/bullet/shape_bullet.cpp]: Ok now it doesn't crash and the user gets a chance to change the physics engine setting in editor

Right now there is a bit of console spam when loading scenes with cylinder collisions nodes and when resizing new cylinder resources

An alternative would be to implement a dummy ShapeSW which does nothing or falls back to a convex collider but dunno... could introduce more problems than it solves

Comment (AndreaCatania) [reply modules/bullet/shape_bullet.cpp]: It's enough if you show a warning only when the cylinder is created

Comment (muiroc) [reply modules/bullet/shape_bullet.cpp]: Sorry, to be clear, this is what the user sees when he creates a cylinder collision shape in GodotPhysics:

cyli_err

As you can see the console outputs the main error message, but also some failed assertions.
Same happens when loading a scene that contains cylinder collision shapes (slightly more asserts tbh)

Is it ok that way?

**Comment by AndreaCatania** [comment]
Hmm, in this case I think it's fine like this.. I will check it better and
I will let you know

**Comment by akien-mga** [comment]
There are some small style issues as shown by this diff: https://travis-ci.org/godotengine/godot/jobs/393317423
See http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html for how to setup style checks locally.

The commits would need to be squashed into one too before merging, see http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html

File: scene/resources/cylinder_shape.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+	ADD_PROPERTY(PropertyInfo(Variant::REAL, "radius", PROPERTY_HINT_RANGE, "0.01,4096,0.01"), "set_radius", "get_radius");
+	ADD_PROPERTY(PropertyInfo(Variant::REAL, "height", PROPERTY_HINT_RANGE, "0.01,4096,0.01"), "set_height", "get_height");
+}
+
+CylinderShape::CylinderShape() :
+		Shape(PhysicsServer::get_singleton()->shape_create(PhysicsServer::SHAPE_CYLINDER)) {
+
+	radius = 1.0;
+	height = 1.0;
```
Comment (AndreaCatania) [inline scene/resources/cylinder_shape.cpp]: set `height = 2.0;` to match the mesh height

**Comment by AndreaCatania** [comment]
@akien-mga when the last change is done can be merged

**Comment by muiroc** [comment]
Cool! Just wanted to say that I'm using this for KinematicBody characters and I really like the resulting  behavior and feeling especially when getting close to ledges, stepping stairs and the resulting normal when snapping to the ground.
Asap when the controller script is cleaned up I will share it (ofc don't expect something perfect!).

**Comment by mhilbrunner** [comment]
Thanks for this contribution!

**Pull request merged**